{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "37782188",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _read_stripes(file_name):\n",
    "    \n",
    "    hold_dict = {x:[] for x in [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]}\n",
    "    source_dict = {5:{x:[] for x in [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]},3:{x:[] for x in [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]}}\n",
    "    \n",
    "    check_c = [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]\n",
    "    var5 = []\n",
    "    var3 = []\n",
    "    incre=-1\n",
    "#     with open(\"/home/spmoran/temp_smoran/Fan_StripeCaller/2024_SummerPublishPush/zebra_merge/GM12878.bedpe\", \"r\") as f:\n",
    "    with open(file_name, \"r\") as f:\n",
    "        f.readline()\n",
    "        for line in f:\n",
    "            chromo,c1,c2,_,c4,c5,pval=line.rstrip(\"\\n\").split()\n",
    "            if chromo != check_c[0]:\n",
    "                check_c.pop(0)\n",
    "            incre+=1\n",
    "            if int(c2)-int(c1)<int(c5)-int(c4):\n",
    "                source_dict[5][chromo]+=[tuple((incre,chromo,*[int(x) for x in [c1,c2,c4,c5]]+[float(pval)]))]\n",
    "            else:\n",
    "                source_dict[3][chromo]+=[tuple((incre,chromo,*[int(x) for x in [c1,c2,c4,c5]]+[float(pval)]))]\n",
    "                \n",
    "    return source_dict, hold_dict\n",
    "\n",
    "def _read_peaks(file_name):\n",
    "    \n",
    "    hold_pdict = {x:[] for x in [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]+[\"chrY\"]}\n",
    "    source_pdict = {x:[] for x in [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]+[\"chrY\"]}\n",
    "    \n",
    "    check_c = [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"] + [\"chrY\"]\n",
    "    var5 = []\n",
    "    var3 = []\n",
    "    incre=-1\n",
    "#     with open(\"/home/spmoran/temp_smoran/Fan_StripeCaller/2024_SummerPublishPush/zebra_merge/GM12878.bedpe\", \"r\") as f:\n",
    "    with open(file_name, \"r\") as f:\n",
    "        f.readline()\n",
    "        for line in f:\n",
    "            chromo,c1,c2,payload=line.split()\n",
    "            if chromo != check_c[0]:\n",
    "                check_c.pop(0)\n",
    "            incre+=1\n",
    "#             if int(c2)-int(c1)<int(c5)-int(c4):\n",
    "            source_pdict[chromo]+=[tuple((incre,chromo,*[int(x) for x in [c1,c2,payload]]))]\n",
    "                \n",
    "    return source_pdict, hold_pdict\n",
    "\n",
    "\n",
    "\n",
    "def IntersectingIntervals(arr1, w_1, arr2, spacer=50000):\n",
    "    i = j = 0\n",
    "    n = len(arr1)\n",
    "    m = len(arr2)\n",
    "        \n",
    "    result=[]\n",
    "    while i < n and j < m:\n",
    "        while (j < m and arr2[j][3]<arr1[i][w_1]):\n",
    "            j+=1\n",
    "            \n",
    "        if j>=m: #one level up?\n",
    "            break\n",
    "        \n",
    "        if arr1[i][w_1]-spacer<=arr2[j][2]<=arr1[i][w_1+1]+spacer and arr1[i][w_1]-spacer<=arr2[j][3]<=arr1[i][w_1+1]+spacer and arr2[j][-1]!=0:\n",
    "            result+=[arr1[i]]\n",
    "            i+=1\n",
    "        elif arr2[j][2]>arr1[i][w_1+1]+spacer:\n",
    "            i+=1\n",
    "        else:\n",
    "            j+=1\n",
    "        \n",
    "    return result\n",
    "\n",
    "\n",
    "def _load_Intersect(source_dict,source_peaks,hold_dict,spacer):\n",
    "    for key in [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]:\n",
    "        hold_dict[key]=IntersectingIntervals(source_dict[5][key],2,source_peaks[key],spacer=2000)\n",
    "        hold_dict[key]+=IntersectingIntervals(source_dict[3][key],4,source_peaks[key],spacer=2000)\n",
    "        \n",
    "def _write_Intersect(hold_dict, filepath):\n",
    "    write_out = \"\"\n",
    "    for key in [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]:\n",
    "        for value in hold_dict[key]:\n",
    "            payload = value[-1]\n",
    "            c5=value[-2]\n",
    "            c4=value[-3]\n",
    "            write_out+=\"\\t\".join([str(y) for y in value[1:4]]+[value[1]]+[str(c4)]+[str(c5)]+[str(payload)])+\"\\n\"\n",
    "            \n",
    "    with open(filepath, \"w\") as outfile:\n",
    "        outfile.write(write_out)\n",
    "        \n",
    "        \n",
    "def _z_write_Intersect(hold_dict, filepath):\n",
    "    write_out = \"\"\n",
    "    for key in [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]:\n",
    "        for value in hold_dict[key]:\n",
    "            svalue = [str(x) for x in value]\n",
    "\n",
    "            write_out+=\"\\t\".join([y for y in svalue[1:4]]+[svalue[1]]+[z for z in svalue[4:]])+\"\\n\"\n",
    "            \n",
    "    with open(filepath, \"w\") as outfile:\n",
    "        outfile.write(write_out)\n",
    "        \n",
    "def _snn_write_Intersect(hold_dict, filepath):\n",
    "    write_out = \"\"\n",
    "    for key in [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]:\n",
    "        for value in hold_dict[key]:\n",
    "            svalue = [str(x) for x in value]\n",
    "            write_out+=\"\\t\".join([y for y in svalue[1:4]]+[svalue[1]]+[z for z in svalue[4:]])+\"\\n\"            \n",
    "    with open(filepath, \"w\") as outfile:\n",
    "        outfile.write(write_out)\n",
    "        \n",
    "def call_CTCF_stripes(stripename, ctcfname, fileoutname,spacer=50000):\n",
    "    sdict, hdict=_read_stripes(stripename)\n",
    "    source_pdict, hold_pdict = _read_peaks(ctcfname)\n",
    "    _load_Intersect(sdict, source_pdict,hdict,spacer)\n",
    "        #noIntersectingIntervals(sdict[5]['chr1'],2,source_pdict['chr1'],spacer=5000))\n",
    "    _write_Intersect(hdict, fileoutname)\n",
    "    \n",
    "def z_call_CTCF_stripes(stripename, ctcfname, fileoutname,spacer=50000):\n",
    "    sdict, hdict=_zebra_read_stripes(stripename)\n",
    "    source_pdict, hold_pdict = _read_peaks(ctcfname)\n",
    "    _load_Intersect(sdict, source_pdict,hdict,spacer)\n",
    "        #noIntersectingIntervals(sdict[5]['chr1'],2,source_pdict['chr1'],spacer=5000))\n",
    "    _z_write_Intersect(hdict, fileoutname)\n",
    "    \n",
    "def snn_call_CTCF_stripes(stripename, ctcfname, fileoutname,spacer=50000):\n",
    "    sdict, hdict=_stripeNN_read_stripes(stripename)\n",
    "    source_pdict, hold_pdict = _read_peaks(ctcfname)\n",
    "    _load_Intersect(sdict, source_pdict,hdict,spacer)\n",
    "        #noIntersectingIntervals(sdict[5]['chr1'],2,source_pdict['chr1'],spacer=5000))\n",
    "    _snn_write_Intersect(hdict, fileoutname)\n",
    "    \n",
    "    \n",
    "def noIntersectingIntervals(arr1, w_1, arr2, spacer=50000):\n",
    "    i = j = 0\n",
    "    n = len(arr1)\n",
    "    m = len(arr2)\n",
    "        \n",
    "    result=[]\n",
    "    while i < n and j < m:\n",
    "        while (j < m and arr2[j][3]<arr1[i][w_1]):\n",
    "            j+=1\n",
    "            \n",
    "        if j>=m: #one level up?\n",
    "            break\n",
    "        \n",
    "        if arr1[i][w_1]-spacer<=arr2[j][2]<=arr1[i][w_1+1]+spacer and arr1[i][w_1]-spacer<=arr2[j][3]<=arr1[i][w_1+1]+spacer and arr2[j][-1]!=0:\n",
    "            i+=1\n",
    "        elif arr2[j][2]>arr1[i][w_1+1]+spacer:\n",
    "            result+=[arr1[i]]\n",
    "            i+=1\n",
    "        else:\n",
    "            j+=1\n",
    "        \n",
    "    return result\n",
    "\n",
    "def _load_noIntersect(source_dict,source_peaks,hold_dict,spacer=2000):\n",
    "    for key in [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]:\n",
    "        hold_dict[key]=noIntersectingIntervals(source_dict[5][key],2,source_peaks[key],spacer)\n",
    "        hold_dict[key]+=noIntersectingIntervals(source_dict[3][key],4,source_peaks[key],spacer)\n",
    "        \n",
    "def _write_noIntersect(hold_dict, filepath):\n",
    "    write_out = \"\"\n",
    "    for key in [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]:\n",
    "        for value in hold_dict[key]:\n",
    "            payload = value[-1]\n",
    "            c5=value[-2]\n",
    "            c4=value[-3]\n",
    "            write_out+=\"\\t\".join([str(y) for y in value[1:4]]+[value[1]]+[str(c4)]+[str(c5)]+[str(payload)])+\"\\n\"\n",
    "            \n",
    "    with open(filepath, \"w\") as outfile:\n",
    "        outfile.write(write_out)\n",
    "\n",
    "def _z_write_noIntersect(hold_dict, filepath):\n",
    "    write_out = \"\"\n",
    "    for key in [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]:\n",
    "        for value in hold_dict[key]:\n",
    "            svalue = [str(x) for x in value]\n",
    "\n",
    "            write_out+=\"\\t\".join([y for y in svalue[1:4]]+[svalue[1]]+[z for z in svalue[4:]])+\"\\n\"\n",
    "            \n",
    "    with open(filepath, \"w\") as outfile:\n",
    "        outfile.write(write_out)\n",
    "\n",
    "def _snn_write_noIntersect(hold_dict, filepath):\n",
    "    write_out = \"\"\n",
    "    for key in [f\"chr{n}\" for n in range(1,23)] + [\"chrX\"]:\n",
    "        for value in hold_dict[key]:\n",
    "            svalue = [str(x) for x in value]\n",
    "            write_out+=\"\\t\".join([y for y in svalue[1:4]]+[svalue[1]]+[z for z in svalue[4:]])+\"\\n\"            \n",
    "    with open(filepath, \"w\") as outfile:\n",
    "        outfile.write(write_out)\n",
    "\n",
    "\n",
    "def call_non_CTCF_stripes(stripename, ctcfname, fileoutname,spacer=1000):\n",
    "    sdict, hdict=_read_stripes(stripename)\n",
    "    source_pdict, hold_pdict = _read_peaks(ctcfname)\n",
    "    _load_noIntersect(sdict, source_pdict,hdict,spacer)\n",
    "        #noIntersectingIntervals(sdict[5]['chr1'],2,source_pdict['chr1'],spacer=5000))\n",
    "    _write_noIntersect(hdict, fileoutname)\n",
    "\n",
    "def z_call_non_CTCF_stripes(stripename, ctcfname, fileoutname,spacer=1000):\n",
    "    sdict, hdict=_zebra_read_stripes(stripename)\n",
    "    source_pdict, hold_pdict = _read_peaks(ctcfname)\n",
    "    _load_noIntersect(sdict, source_pdict,hdict,spacer)\n",
    "        #noIntersectingIntervals(sdict[5]['chr1'],2,source_pdict['chr1'],spacer=5000))\n",
    "    _z_write_noIntersect(hdict, fileoutname)\n",
    "    \n",
    "def snn_call_non_CTCF_stripes(stripename, ctcfname, fileoutname,spacer=1000):\n",
    "    sdict, hdict=_stripeNN_read_stripes(stripename)\n",
    "    source_pdict, hold_pdict = _read_peaks(ctcfname)\n",
    "    _load_noIntersect(sdict, source_pdict,hdict,spacer)\n",
    "        #noIntersectingIntervals(sdict[5]['chr1'],2,source_pdict['chr1'],spacer=5000))\n",
    "    _snn_write_noIntersect(hdict, fileoutname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "90df9f2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_intervals_by_chromosome(file_path):\n",
    "    \"\"\"\n",
    "    Reads intervals from File A and organizes them in a dictionary keyed by chromosome.\n",
    "    \"\"\"\n",
    "    chrom_intervals = {}\n",
    "    with open(file_path, 'r') as f:\n",
    "        for line in f:\n",
    "            parts = line.strip().split()\n",
    "            if len(parts) >= 3:\n",
    "                chrom = parts[0]\n",
    "                interval = (parts[1], parts[2])  # (start, end) as strings\n",
    "                if chrom not in chrom_intervals:\n",
    "                    chrom_intervals[chrom] = set()\n",
    "                chrom_intervals[chrom].add(interval)\n",
    "    return chrom_intervals\n",
    "\n",
    "def find_matching_loops_by_chromosome(file_a_path, file_b_path, output_path=None):\n",
    "    \"\"\"\n",
    "    Finds and prints (or saves) loops from File B that match any interval from File A.\n",
    "    Matching is chromosome-aware.\n",
    "    \"\"\"\n",
    "    intervals_by_chr = read_intervals_by_chromosome(file_a_path)\n",
    "\n",
    "    with open(file_b_path, 'r') as file_b:\n",
    "        output_lines = []\n",
    "        for line in file_b:\n",
    "            parts = line.strip().split()\n",
    "            if len(parts) >= 6:\n",
    "                chrom1, start1, end1 = parts[0:3]\n",
    "                chrom2, start2, end2 = parts[3:6]\n",
    "                \n",
    "                match1 = chrom1 in intervals_by_chr and (start1, end1) in intervals_by_chr[chrom1]\n",
    "                match2 = chrom2 in intervals_by_chr and (start2, end2) in intervals_by_chr[chrom2]\n",
    "                \n",
    "                if match1 or match2:\n",
    "                    output_lines.append(line)\n",
    "\n",
    "    # Output results\n",
    "    if output_path:\n",
    "        with open(output_path, 'w') as out_f:\n",
    "            out_f.writelines(output_lines)\n",
    "    else:\n",
    "        for line in output_lines:\n",
    "            print(line.strip())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32982779",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "prefix=\"./data_files/10KQuaggaCalls/\"\n",
    "sourcefile=\"10000nt_MR10000000_ML50000_MD2000000_MW80000_WS2_SG1RH0.1_NSTR2_WholeGenome_hg38_GM12878_width_filtered.txt\"\n",
    "zebra='../stripe_results/GM12878.bedpe'\n",
    "stripenn='../stripe_results/results_filtered.tsv'\n",
    "\n",
    "awk '$3 - $2 <= $6 - $5 { OFS = \"\\t\"; print $1, $2, $3, $4, $5, $6, $7 }' ${prefix}${sourcefile} > ./uniqueAnalysis/v_or_h/q_5prime.txt\n",
    "awk '$3 - $2 > $6 - $5 { OFS = \"\\t\"; print $1, $2, $3, $4, $5, $6, $7 }' ${prefix}${sourcefile} > ./uniqueAnalysis/v_or_h/q_3prime.txt\n",
    "\n",
    "awk '$3 - $2 <= $6 - $5 { OFS = \"\\t\"; print $1, $2, $3, $4, $5, $6, $7 }' ${zebra} > ./uniqueAnalysis/v_or_h/z_5prime.txt\n",
    "awk '$3 - $2 > $6 - $5 { OFS = \"\\t\"; print $1, $2, $3, $4, $5, $6, $7 }' ${zebra} > ./uniqueAnalysis/v_or_h/z_3prime.txt\n",
    "\n",
    "# stripenn has different\n",
    "awk -F'\\t' '{ if (int($2) == int($5)) print $1, $2, $3, $4, $5, $6, $7 }' ${stripenn} > ./uniqueAnalysis/v_or_h/s_5prime.txt\n",
    "awk -F'\\t' '{ if (int($2) != int($5)) print $1, $2, $3, $4, $5, $6, $7 }' ${stripenn} > ./uniqueAnalysis/v_or_h/s_3prime.txt\n",
    "\n",
    "awk '{ OFS = \"\\t\"; print $1, $2, $3 }' ./uniqueAnalysis/v_or_h/q_5prime.txt > ./uniqueAnalysis/v_or_h/q_5prime_cut.txt\n",
    "awk '{ OFS = \"\\t\"; print $1, $2, $3 }' ./uniqueAnalysis/v_or_h/z_5prime.txt > ./uniqueAnalysis/v_or_h/z_5prime_cut.txt\n",
    "awk '{ OFS = \"\\t\"; print $1, $2, $3 }' ./uniqueAnalysis/v_or_h/s_5prime.txt > ./uniqueAnalysis/v_or_h/s_5prime_cut.txt\n",
    "\n",
    "awk '{ OFS = \"\\t\"; print $4, $5, $6 }' ./uniqueAnalysis/v_or_h/q_3prime.txt > ./uniqueAnalysis/v_or_h/q_3prime_cut.txt\n",
    "awk '{ OFS = \"\\t\"; print $4, $5, $6 }' ./uniqueAnalysis/v_or_h/z_3prime.txt > ./uniqueAnalysis/v_or_h/z_3prime_cut.txt\n",
    "awk '{ OFS = \"\\t\"; print $4, $5, $6 }' ./uniqueAnalysis/v_or_h/s_3prime.txt > ./uniqueAnalysis/v_or_h/s_3prime_cut.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d67f6378",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# use uniq, as many can fit within larger stripes.\n",
    "mkdir unqiueAnalysis\n",
    "mkdir ./unqiueAnalysis/v_or_h\n",
    "mkdir ./unqiueAnalysis/v_or_h/unique\n",
    "mkdir ./unqiueAnalysis/finalout\n",
    "\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/q_5prime_cut.txt -b ./unqiueAnalysis/v_or_h/z_5prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/unique/q_no_z_5prime_cut.txt\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/q_3prime_cut.txt -b ./unqiueAnalysis/v_or_h/z_3prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/unique/q_no_z_3prime_cut.txt\n",
    "\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/q_5prime_cut.txt -b ./unqiueAnalysis/v_or_h/s_5prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/unique/q_no_s_5prime_cut.txt\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/q_3prime_cut.txt -b ./unqiueAnalysis/v_or_h/s_3prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/unique/q_no_s_3prime_cut.txt\n",
    "\n",
    "\n",
    "\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/z_5prime_cut.txt -b ./unqiueAnalysis/v_or_h/q_5prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/unique/z_no_q_5prime_cut.txt\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/z_3prime_cut.txt -b ./unqiueAnalysis/v_or_h/q_3prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/unique/z_no_q_3prime_cut.txt\n",
    "\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/z_5prime_cut.txt -b ./unqiueAnalysis/v_or_h/s_5prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/unique/z_no_s_5prime_cut.txt\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/z_3prime_cut.txt -b ./unqiueAnalysis/v_or_h/s_3prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/unique/z_no_s_3prime_cut.txt\n",
    "\n",
    "\n",
    "\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/s_5prime_cut.txt -b ./unqiueAnalysis/v_or_h/q_5prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/unique/s_no_q_5prime_cut.txt\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/s_3prime_cut.txt -b ./unqiueAnalysis/v_or_h/q_3prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/unique/s_no_q_3prime_cut.txt\n",
    "\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/s_5prime_cut.txt -b ./unqiueAnalysis/v_or_h/z_5prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/unique/s_no_z_5prime_cut.txt\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/s_3prime_cut.txt -b ./unqiueAnalysis/v_or_h/z_3prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/unique/s_no_z_3prime_cut.txt\n",
    "\n",
    "\n",
    "\n",
    "# Intersection between all tools, perspective of q_*, s_*, or z_*\n",
    "\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/intersects/q_no_z_5prime_cut.txt -b ./unqiueAnalysis/v_or_h/s_5prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/intersects/q_no_zs_5prime_cut.txt\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/intersects/q_no_z_3prime_cut.txt -b ./unqiueAnalysis/v_or_h/s_3prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/intersects/q_no_zs_3prime_cut.txt\n",
    "\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/intersects/s_no_z_5prime_cut.txt -b ./unqiueAnalysis/v_or_h/q_5prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/intersects/s_no_zq_5prime_cut.txt\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/intersects/s_no_z_3prime_cut.txt -b ./unqiueAnalysis/v_or_h/q_3prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/intersects/s_no_zq_3prime_cut.txt\n",
    "\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/intersects/z_no_q_5prime_cut.txt -b ./unqiueAnalysis/v_or_h/s_5prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/intersects/z_no_qs_5prime_cut.txt\n",
    "bedtools intersect -v -a ./unqiueAnalysis/v_or_h/intersects/z_no_q_3prime_cut.txt -b ./unqiueAnalysis/v_or_h/s_3prime_cut.txt| uniq  > ./unqiueAnalysis/v_or_h/intersects/z_no_qs_3prime_cut.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74b4b37c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "\n",
    "# #may need to remove headers from final files below\n",
    "\n",
    "# cat ./unqiueAnalysis/v_or_h/intersects/q_no_zs_5prime_cut.txt ./unqiueAnalysis/v_or_h/intersects/q_no_zs_3prime_cut.txt > ./unqiueAnalysis/finalout/q_no_zs_final_cut.txt\n",
    "# cat ./unqiueAnalysis/v_or_h/intersects/s_no_zq_5prime_cut.txt ./unqiueAnalysis/v_or_h/intersects/s_no_zq_3prime_cut.txt > ./unqiueAnalysis/finalout/s_no_zq_final_cut.txt\n",
    "# cat ./unqiueAnalysis/v_or_h/intersects/z_no_qs_5prime_cut.txt ./unqiueAnalysis/v_or_h/intersects/z_no_qs_3prime_cut.txt > ./unqiueAnalysis/finalout/z_no_qs_final_cut.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "152de6bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir ./unqiueAnalysis/remapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "47e463ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "stripepath={\n",
    "    'Quagga': \"./unqiueAnalysis/finalout/q_no_zs_final_cut.txt\",\n",
    "    'Zebra': \"./unqiueAnalysis/finalout/s_no_zq_final_cut.txt\",\n",
    "    'StripeNN': \"./unqiueAnalysis/finalout/z_no_qs_final_cut.txt\"\n",
    "}\n",
    "\n",
    "pathA = \"./unqiueAnalysis/v_or_h/q_3prime.txt\"\n",
    "pathB = \"./unqiueAnalysis/v_or_h/z_3prime.txt\"\n",
    "pathC = \"./unqiueAnalysis/v_or_h/s_3prime.txt\"\n",
    "\n",
    "find_matching_loops_by_chromosome(stripepath['Quagga'], pathA, \"./unqiueAnalysis/remapped/\"+\"quagga_uniq_3prime_remapped.txt\")\n",
    "find_matching_loops_by_chromosome(stripepath['Zebra'], pathB, \"./unqiueAnalysis/remapped/\"+\"zebra_uniq_3prime_remapped.txt\")\n",
    "find_matching_loops_by_chromosome(stripepath['StripeNN'], pathC, \"./unqiueAnalysis/remapped/\"+\"stripenn_uniq_3prime_remapped.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58a8a31d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "###!!!! fuse 5 and 3remapped and make into new epigenomic pileups and APA.\n",
    "# may need to remove headers manually \"#chr x1 x2, or chr pos1 pos2, etc.\"\n",
    "\n",
    "cat ./unqiueAnalysis/v_or_h/q_5prime.txt ./unqiueAnalysis/remapped/quagga_uniq_3prime_remapped.txt > ./unqiueAnalysis/finalout/q_no_zs_final_cut.txt\n",
    "cat ./unqiueAnalysis/v_or_h/s_5prime.txt ./unqiueAnalysis/remapped/stripenn_uniq_3prime_remapped.txt > ./unqiueAnalysis/finalout/s_no_zq_final_cut.txt\n",
    "cat ./unqiueAnalysis/v_or_h/z_5prime.txt ./unqiueAnalysis/remapped/zebra_uniq_3prime_remapped.txt > ./unqiueAnalysis/finalout/z_no_qs_final_cut.txt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
